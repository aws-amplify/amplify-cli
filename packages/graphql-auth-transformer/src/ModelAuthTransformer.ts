import { Transformer, TransformerContext, InvalidDirectiveError } from 'graphql-transformer-core'
import GraphQLAPI from 'cloudform/types/appSync/graphQlApi'
import { ResourceFactory } from './resources'
import { AuthRule, ModelQuery, ModelMutation } from './AuthRule'
import { ObjectTypeDefinitionNode, DirectiveNode, ArgumentNode } from 'graphql'
import { ResourceConstants, ResolverResourceIDs, isListType } from 'graphql-transformer-common'
import {
    Expression, print, raw, iff, equals, forEach, set, ref, list, compoundExpression, or, newline,
    comment
} from 'graphql-mapping-template';
import { valueFromASTUntyped } from 'graphql'

import {
    OWNER_AUTH_STRATEGY,
    DEFAULT_OWNER_FIELD,
    DEFAULT_IDENTITY_FIELD,
    GROUPS_AUTH_STRATEGY,
    DEFAULT_GROUPS_FIELD
} from './constants'

/**
 * Implements the ModelAuthTransformer.
 *
 * Owner Auth Usage:
 *
 * type Post @auth(allow: owner) {
 *   id: ID!
 *   title: String
 *   createdAt: String
 *   updatedAt: String
 * }
 *
 * Impact:
 *
 * getPost - In the response mapping template we check the "owner" field === $ctx.identity.username.
 * listPost - In the response mapping template we return only items where "owner" === $ctx.identity.username
 * createPost - We automatically insert a "owner" field to attribute values where "owner" === $ctx.identity.username.
 * updatePost - Expose "owner" field in input/output and would set conditional update expression to look for owner.
 * deletePost - Conditional expression checking that the owner === $ctx.identity.username
 *
 * Note: The name of the "owner" field may be configured via the CF paramaters.
 *
 * type Post @auth(allow: groups, groups: ["Admin", "Dev"]) {
 *   id: ID!
 *   title: String
 *   createdAt: String
 *   updatedAt: String
 * }
 *
 * Impact:
 *
 * getPost - Update req template to look for the groups in the identity.
 * listPost - Update req template to look for the groups in the identity.
 * createPost - Update req template to look for the groups in the identity.
 * updatePost - Update req template to look for the groups in the identity.
 * deletePost - Update req template to look for the groups in the identity.
 *
 * TODO: Document support for dynamic group authorization against
 * attributes of the records using conditional expressions. This will likely
 * be via a new argument such as "groupsField".
 */
export class ModelAuthTransformer extends Transformer {

    resources: ResourceFactory

    constructor() {
        super(
            'ModelAuthTransformer',
            `
            directive @auth(rules: [AuthRule!]!) on OBJECT
            input AuthRule {
                allow: AuthStrategy!
                ownerField: String # defaults to "owner"
                identityField: String # defaults to "username"
                groupsField: String
                groups: [String]
                queries: [ModelQuery]
                mutations: [ModelMutation]
            }
            enum AuthStrategy { owner groups }
            enum ModelQuery { get list }
            enum ModelMutation { create update delete }
            `
        )
        this.resources = new ResourceFactory();
    }

    /**
     * Updates the GraphQL API record to use user pool auth.
     */
    private updateAPIForUserPools = (ctx: TransformerContext): void => {
        const apiRecord = ctx.getResource(ResourceConstants.RESOURCES.GraphQLAPILogicalID) as GraphQLAPI
        const updated = this.resources.updateGraphQLAPIWithAuth(apiRecord)
        ctx.setResource(ResourceConstants.RESOURCES.GraphQLAPILogicalID, updated)
    }

    public before = (ctx: TransformerContext): void => {
        const template = this.resources.initTemplate();
        ctx.mergeResources(template.Resources)
        ctx.mergeParameters(template.Parameters)
        ctx.mergeOutputs(template.Outputs)
        ctx.mergeConditions(template.Conditions)
        this.updateAPIForUserPools(ctx)
    }

    /**
     * Implement the transform for an object type. Depending on which operations are to be protected
     */
    public object = (def: ObjectTypeDefinitionNode, directive: DirectiveNode, ctx: TransformerContext): void => {
        const get = (s: string) => (arg: ArgumentNode) => arg.name.value === s
        const getArg = (arg: string, dflt?: any) => {
            const argument = directive.arguments.find(get(arg))
            return argument ? valueFromASTUntyped(argument.value) : dflt
        }

        const modelDirective = def.directives.find((dir) => dir.name.value === 'model')
        if (!modelDirective) {
            throw new InvalidDirectiveError('Types annotated with @auth must also be annotated with @model.')
        }

        // Get the auth rules.
        const rules = getArg('rules', []) as AuthRule[]

        // Create a reverse index on rules from operation -> rules list.
        const queryRules: { [k in ModelQuery]: AuthRule[] } = {
            get: [],
            list: [],
        }
        const mutationRules: { [k in ModelMutation]: AuthRule[] } = {
            create: [],
            update: [],
            delete: []
        }
        const matchQuery = (op: ModelQuery) => (rule: AuthRule) => {
            if (rule.queries) {
                const matchesOp = rule.queries.find(o => o === op)
                return Boolean(matchesOp)
            } else if (rule.queries === null) {
                return false
            }
            return true
        }
        const matchMutation = (op: ModelMutation) => (rule: AuthRule) => {
            if (rule.mutations) {
                const matchesOp = rule.mutations.find(o => o === op)
                return Boolean(matchesOp)
            } else if (rule.mutations === null) {
                return false
            }
            return true
        }
        for (const rule of rules) {
            if (matchQuery('get')(rule)) {
                queryRules.get.push(rule)
            }
            if (matchQuery('list')(rule)) {
                queryRules.list.push(rule)
            }
            if (matchMutation('create')(rule)) {
                mutationRules.create.push(rule)
            }
            if (matchMutation('update')(rule)) {
                mutationRules.update.push(rule)
            }
            if (matchMutation('delete')(rule)) {
                mutationRules.delete.push(rule)
            }
        }

        // For each operation evaluate the rules and apply the changes to the relevant resolver.
        this.protectCreateMutation(ctx, ResolverResourceIDs.DynamoDBCreateResolverResourceID(def.name.value), mutationRules.create, def)
        this.protectUpdateMutation(ctx, ResolverResourceIDs.DynamoDBUpdateResolverResourceID(def.name.value), mutationRules.update, def)
        this.protectDeleteMutation(ctx, ResolverResourceIDs.DynamoDBDeleteResolverResourceID(def.name.value), mutationRules.delete, def)
        this.protectGetQuery(ctx, ResolverResourceIDs.DynamoDBGetResolverResourceID(def.name.value), queryRules.get)
        this.protectListQuery(ctx, ResolverResourceIDs.DynamoDBListResolverResourceID(def.name.value), queryRules.list)
    }

    /**
     * Protect get queries.
     * If static group:
     *  If statically authorized then allow the operation. Stop.
     * If owner and/or dynamic group:
     *  If the result item satisfies the owner/group authorization condition
     *  then allow it.
     * @param ctx The transformer context.
     * @param resolverResourceId The logical id of the get resolver.
     * @param rules The auth rules to apply.
     */
    private protectGetQuery(ctx: TransformerContext, resolverResourceId: string, rules: AuthRule[]) {
        const resolver = ctx.getResource(resolverResourceId)
        if (!rules || rules.length === 0 || !resolver) {
            return
        } else {

            // Break the rules out by strategy.
            const staticGroupAuthorizationRules = this.getStaticGroupRules(rules)
            const dynamicGroupAuthorizationRules = this.getDynamicGroupRules(rules)
            const ownerAuthorizationRules = this.getOwnerRules(rules)

            // Generate the expressions to validate each strategy.
            const staticGroupAuthorizationExpression = this.resources.staticGroupAuthorizationExpression(staticGroupAuthorizationRules)
            const dynamicGroupAuthorizationExpression = this.resources.dynamicGroupAuthorizationExpressionForReadOperations(
                dynamicGroupAuthorizationRules
            )
            const ownerAuthorizationExpression = this.resources.ownerAuthorizationExpressionForReadOperations(
                ownerAuthorizationRules
            )
            const throwIfUnauthorizedExpression = this.resources.throwIfUnauthorized()

            // Update the existing resolver with the authorization checks.
            const responseMappingTemplatePrefixExpressions = [
                staticGroupAuthorizationExpression,
                newline(),
                dynamicGroupAuthorizationExpression,
                newline(),
                ownerAuthorizationExpression,
                newline(),
                throwIfUnauthorizedExpression
            ]
            const templateParts = [
                print(compoundExpression(responseMappingTemplatePrefixExpressions)),
                resolver.Properties.ResponseMappingTemplate
            ]
            resolver.Properties.ResponseMappingTemplate = templateParts.join('\n\n')
            ctx.setResource(resolverResourceId, resolver)
        }
    }

    /**
     * Protect list queries.
     * If static group:
     *  If the user is statically authorized then return items and stop.
     * If dynamic group and/or owner:
     *  Loop through all items and find items that satisfy any of the group or
     *  owner conditions.
     * @param ctx The transformer context.
     * @param resolverResourceId The logical id of the resolver to be updated in the CF template.
     * @param rules The set of rules that apply to the operation.
     */
    private protectListQuery(ctx: TransformerContext, resolverResourceId: string, rules: AuthRule[]) {
        const resolver = ctx.getResource(resolverResourceId)
        if (!rules || rules.length === 0 || !resolver) {
            return
        } else {
            // Break the rules out by strategy.
            const staticGroupAuthorizationRules = this.getStaticGroupRules(rules)
            const dynamicGroupAuthorizationRules = this.getDynamicGroupRules(rules)
            const ownerAuthorizationRules = this.getOwnerRules(rules)

            // Generate the expressions to validate each strategy.
            const staticGroupAuthorizationExpression = this.resources.staticGroupAuthorizationExpression(staticGroupAuthorizationRules)

            // In list queries, the dynamic group and ownership authorization checks
            // occur on a per item basis. The helpers take the variable names
            // as parameters to allow for this use case.
            const dynamicGroupAuthorizationExpression = this.resources.dynamicGroupAuthorizationExpressionForReadOperations(
                dynamicGroupAuthorizationRules,
                'item',
                ResourceConstants.SNIPPETS.IsLocalDynamicGroupAuthorizedVariable,
                raw(`false`)
            )
            const ownerAuthorizationExpression = this.resources.ownerAuthorizationExpressionForReadOperations(
                ownerAuthorizationRules,
                'item',
                ResourceConstants.SNIPPETS.IsLocalOwnerAuthorizedVariable,
                raw(`false`)
            )
            const appendIfLocallyAuthorized = this.resources.appendItemIfLocallyAuthorized()

            const ifNotStaticallyAuthedFilterObjects = iff(
                raw(`! $${ResourceConstants.SNIPPETS.IsStaticGroupAuthorizedVariable}`),
                compoundExpression([
                    set(ref('items'), list([])),
                    forEach(
                        ref('item'),
                        ref('ctx.result.items'),
                        [
                            dynamicGroupAuthorizationExpression,
                            newline(),
                            ownerAuthorizationExpression,
                            newline(),
                            appendIfLocallyAuthorized
                        ]
                    ),
                    set(ref('ctx.result.items'), ref('items'))
                ])
            )
            const templateParts = [
                print(
                    compoundExpression([
                        staticGroupAuthorizationExpression,
                        newline(),
                        comment('[Start] If not static group authorized, filter items'),
                        ifNotStaticallyAuthedFilterObjects,
                        comment('[End] If not static group authorized, filter items')
                    ])
                ),
                resolver.Properties.ResponseMappingTemplate
            ]
            resolver.Properties.ResponseMappingTemplate = templateParts.join('\n\n')
            ctx.setResource(resolverResourceId, resolver)
        }
    }

    /**
     * Inject auth rules for create mutations.
     * If owner auth:
     *  If the owner field exists in the input, validate that it against the identity.
     *  If the owner field dne in the input, insert the identity.
     * If group:
     *  If the user is static group authorized allow operation no matter what.
     *  If dynamic group and the input defines a group(s) validate it against the identity.
     * @param ctx
     * @param resolverResourceId
     * @param rules
     */
    private protectCreateMutation(
        ctx: TransformerContext,
        resolverResourceId: string,
        rules: AuthRule[],
        parent: ObjectTypeDefinitionNode
    ) {
        const resolver = ctx.getResource(resolverResourceId)
        if (!rules || rules.length === 0 || !resolver) {
            return
        } else {
            // Break the rules out by strategy.
            const staticGroupAuthorizationRules = this.getStaticGroupRules(rules)
            const dynamicGroupAuthorizationRules = this.getDynamicGroupRules(rules)
            const ownerAuthorizationRules = this.getOwnerRules(rules)

            // Generate the expressions to validate each strategy.
            const staticGroupAuthorizationExpression = this.resources.staticGroupAuthorizationExpression(staticGroupAuthorizationRules)

            // In create mutations, the dynamic group and ownership authorization checks
            // are done before calling PutItem.
            const dynamicGroupAuthorizationExpression = this.resources.dynamicGroupAuthorizationExpressionForCreateOperations(
                dynamicGroupAuthorizationRules
            )
            const fieldIsList = (fieldName: string) => {
                const field = parent.fields.find(field => field.name.value === fieldName);
                if (field) {
                    return isListType(field.type);
                }
                return false;
            }
            const ownerAuthorizationExpression = this.resources.ownerAuthorizationExpressionForCreateOperations(
                ownerAuthorizationRules,
                fieldIsList
            )

            const throwIfUnauthorizedExpression = this.resources.throwIfUnauthorized()
            const templateParts = [
                print(
                    compoundExpression([
                        staticGroupAuthorizationExpression,
                        newline(),
                        dynamicGroupAuthorizationExpression,
                        newline(),
                        ownerAuthorizationExpression,
                        newline(),
                        throwIfUnauthorizedExpression
                    ])
                ),
                resolver.Properties.RequestMappingTemplate
            ]
            resolver.Properties.RequestMappingTemplate = templateParts.join('\n\n')
            ctx.setResource(resolverResourceId, resolver)
        }
    }

    /**
     * Protect update and delete mutations.
     * If Owner:
     *  Update the conditional expression such that the update only works if
     *  the user is the owner.
     * If dynamic group:
     *  Update the conditional expression such that it succeeds if the user is
     *  dynamic group authorized. If the operation is also owner authorized this
     *  should be joined with an OR expression.
     * If static group:
     *  If the user is statically authorized then allow no matter what. This can
     *  be done by removing the conditional expression as long as static group
     *  auth is always checked last.
     * @param ctx The transformer context.
     * @param resolverResourceId The logical id of the resolver in the template.
     * @param rules The list of rules to apply.
     */
    private protectUpdateOrDeleteMutation(
        ctx: TransformerContext,
        resolverResourceId: string,
        rules: AuthRule[],
        parent: ObjectTypeDefinitionNode
    ) {
        const resolver = ctx.getResource(resolverResourceId)
        if (!rules || rules.length === 0 || !resolver) {
            return
        } else {
            // Break the rules out by strategy.
            const staticGroupAuthorizationRules = this.getStaticGroupRules(rules)
            const dynamicGroupAuthorizationRules = this.getDynamicGroupRules(rules)
            const ownerAuthorizationRules = this.getOwnerRules(rules)

            // Generate the expressions to validate each strategy.
            const staticGroupAuthorizationExpression = this.resources.staticGroupAuthorizationExpression(staticGroupAuthorizationRules)

            // In create mutations, the dynamic group and ownership authorization checks
            // are done before calling PutItem.
            const dynamicGroupAuthorizationExpression = this.resources.dynamicGroupAuthorizationExpressionForUpdateOrDeleteOperations(
                dynamicGroupAuthorizationRules
            )

            const fieldIsList = (fieldName: string) => {
                const field = parent.fields.find(field => field.name.value === fieldName);
                if (field) {
                    return isListType(field.type);
                }
                return false;
            }
            const ownerAuthorizationExpression = this.resources.ownerAuthorizationExpressionForUpdateOrDeleteOperations(
                ownerAuthorizationRules,
                fieldIsList
            )

            const collectAuthCondition = this.resources.collectAuthCondition()
            const ifNotStaticallyAuthedCreateAuthCondition = iff(
                raw(`! $${ResourceConstants.SNIPPETS.IsStaticGroupAuthorizedVariable}`),
                compoundExpression([
                    dynamicGroupAuthorizationExpression,
                    newline(),
                    ownerAuthorizationExpression,
                    newline(),
                    collectAuthCondition
                ])
            )

            const throwIfNotStaticGroupAuthorizedOrAuthConditionIsEmpty = this.resources.throwIfNotStaticGroupAuthorizedOrAuthConditionIsEmpty()

            const templateParts = [
                print(
                    compoundExpression([
                        staticGroupAuthorizationExpression,
                        newline(),
                        ifNotStaticallyAuthedCreateAuthCondition,
                        newline(),
                        throwIfNotStaticGroupAuthorizedOrAuthConditionIsEmpty
                    ])
                ),
                resolver.Properties.RequestMappingTemplate
            ]
            resolver.Properties.RequestMappingTemplate = templateParts.join('\n\n')
            ctx.setResource(resolverResourceId, resolver)
        }
    }

    private protectUpdateMutation(ctx: TransformerContext, resolverResourceId: string, rules: AuthRule[], parent: ObjectTypeDefinitionNode) {
        return this.protectUpdateOrDeleteMutation(ctx, resolverResourceId, rules, parent)
    }

    private protectDeleteMutation(ctx: TransformerContext, resolverResourceId: string, rules: AuthRule[], parent: ObjectTypeDefinitionNode) {
        return this.protectUpdateOrDeleteMutation(ctx, resolverResourceId, rules, parent)
    }

    private getOwnerRules(rules: AuthRule[]): AuthRule[] {
        return rules.filter(rule => rule.allow === 'owner');
    }

    private getStaticGroupRules(rules: AuthRule[]): AuthRule[] {
        return rules.filter(rule => rule.allow === 'groups' && Boolean(rule.groups));
    }

    private getDynamicGroupRules(rules: AuthRule[]): AuthRule[] {
        return rules.filter(rule => rule.allow === 'groups' && !Boolean(rule.groups));
    }

}
