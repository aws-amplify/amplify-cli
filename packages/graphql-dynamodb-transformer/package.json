{
  "name": "graphql-dynamodb-transformer",
  "version": "6.19.5",
  "description": "An AppSync model transform that takes a simple model and creates a DynamoDB table, DynamoDB stream.",
  "repository": {
    "type": "git",
    "url": "https://github.com/aws-amplify/amplify-cli.git",
    "directory": "packages/graphql-dynamodb-transformer"
  },
  "author": "Amazon Web Services",
  "license": "Apache-2.0",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "keywords": [
    "graphql",
    "appsync",
    "aws"
  ],
  "scripts": {
    "test": "jest",
    "test-ci": "jest --ci -i",
    "build": "tsc",
    "clean": "rimraf ./lib",
    "watch": "tsc -w"
  },
  "dependencies": {
    "@types/pluralize": "^0.0.29",
    "cloudform-types": "^4.2.0",
    "graphql": "^14.5.8",
    "graphql-mapping-template": "4.13.4",
    "graphql-transformer-common": "4.17.4",
    "graphql-transformer-core": "6.19.4",
    "md5": "^2.2.1",
    "pluralize": "^8.0.0"
  },
  "devDependencies": {
    "@types/md5": "^2.1.33",
    "@types/node": "^10.17.13"
  },
  "jest": {
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    },
    "testURL": "http://localhost",
    "testRegex": "(src/__tests__/.*.test.ts)$",
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "jsx",
      "json",
      "node"
    ],
    "collectCoverage": true
  }
}
