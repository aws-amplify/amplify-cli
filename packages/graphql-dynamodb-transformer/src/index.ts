export * from './DynamoDBModelTransformer';
export * from './definitions';
export * from './ModelDirectiveArgs';
