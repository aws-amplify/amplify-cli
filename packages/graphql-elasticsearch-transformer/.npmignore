**/__mocks__/**
**/__tests__/**
src
streaming-lambda
tsconfig.json
tsconfig.tsbuildinfo
