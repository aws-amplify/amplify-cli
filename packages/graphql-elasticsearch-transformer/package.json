{
  "name": "graphql-elasticsearch-transformer",
  "version": "1.0.28",
  "description": "An AppSync model transform that takes a simple model and creates a DynamoDB table, DynamoDB stream, and ES index with the queries to match.",
  "main": "lib/index.js",
  "scripts": {
    "build": "tsc && cd streaming-lambda && bestzip ../lib/streaming-lambda.zip python_streaming_function.py",
    "clean": "rm -rf ./lib"
  },
  "keywords": [
    "graphql",
    "appsync",
    "aws"
  ],
  "author": "Amazon Web Services",
  "license": "Apache-2.0",
  "dependencies": {
    "cloudform": "^2.2.1",
    "graphql": "^0.13.2",
    "graphql-appsync-transformer": "^1.0.28",
    "graphql-mapping-template": "^1.0.28",
    "graphql-transformer-common": "^1.0.28",
    "graphql-transformer-core": "^1.0.28"
  },
  "devDependencies": {
    "@types/jest": "^23.1.1",
    "@types/node": "^10.3.4",
    "bestzip": "^2.0.0",
    "graphql-dynamodb-transformer": "^1.0.28",
    "jest": "^23.1.0",
    "ts-jest": "^22.4.6",
    "tslint": "^5.10.0",
    "typescript": "^2.8.3"
  },
  "jest": {
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    },
    "testURL": "http://localhost",
    "testRegex": "(src/__tests__/.*.test.*)$",
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "jsx",
      "json",
      "node"
    ]
  }
}
