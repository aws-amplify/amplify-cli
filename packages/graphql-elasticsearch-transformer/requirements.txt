boto3==1.7.74
botocore==1.10.74