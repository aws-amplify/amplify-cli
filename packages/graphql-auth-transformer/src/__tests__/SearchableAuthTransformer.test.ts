import { GraphQLTransform } from 'graphql-transformer-core';
import { DynamoDBModelTransformer } from 'graphql-dynamodb-transformer';
import { ModelAuthTransformer } from '../ModelAuthTransformer';
import { SearchableModelTransformer } from 'graphql-elasticsearch-transformer';
const featureFlags = {
  getBoolean: jest.fn().mockImplementation((name, defaultValue) => {
    if (name === 'improvePluralization') {
      return true;
    }
    return;
  }),
  getNumber: jest.fn(),
  getObject: jest.fn(),
  getString: jest.fn(),
};

test('test auth logic is enabled on owner/static rules in resposne es resolver', () => {
  const validSchema = `
        type Comment @model
            @searchable
            @auth(rules: [
                { allow: owner }
                { allow: groups, groups: ["writer"]}
            ])
        {
            id: ID!
            content: String
        }
    `;
  const transformer = new GraphQLTransform({
    featureFlags,
    transformers: [
      new DynamoDBModelTransformer(),
      new SearchableModelTransformer(),
      new ModelAuthTransformer({
        authConfig: {
          defaultAuthentication: {
            authenticationType: 'AMAZON_COGNITO_USER_POOLS',
          },
          additionalAuthenticationProviders: [],
        },
      }),
    ],
  });

  const out = transformer.transform(validSchema);
  // expect response resolver to contain auth logic for owner rule
  expect(out).toBeDefined();
  expect(out.resolvers['Query.searchComments.res.vtl']).toContain(
    '## Authorization rule: { allow: owner, ownerField: "owner", identityClaim: "cognito:username" } **',
  );
  // expect response resolver to contain auth logic for group rule
  expect(out.resolvers['Query.searchComments.res.vtl']).toContain(
    '## Authorization rule: { allow: groups, groups: ["writer"], groupClaim: "cognito:groups" } **',
  );
});

test('test auth logic is enabled for iam/apiKey auth rules in response es resolver', () => {
  const validSchema = `
        type Post @model
            @searchable
            @auth(rules: [
                { allow: public, provider: apiKey } # api key is allowed
                { allow: private, provider: iam } # auth roles are allowed
            ]) {
            id: ID!
            content: String
            secret: String @auth(rules: [{ allow: private, provider: iam }]) # only auth role can do crud on this
        }
    `;
  const transformer = new GraphQLTransform({
    featureFlags,
    transformers: [
      new DynamoDBModelTransformer(),
      new SearchableModelTransformer(),
      new ModelAuthTransformer({
        authConfig: {
          defaultAuthentication: {
            authenticationType: 'AMAZON_COGNITO_USER_POOLS',
          },
          additionalAuthenticationProviders: [
            {
              authenticationType: 'API_KEY',
              apiKeyConfig: {
                description: 'E2E Test API Key',
                apiKeyExpirationDays: 300,
              },
            },
            {
              authenticationType: 'AWS_IAM',
            },
          ],
        },
      }),
    ],
  });

  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.schema).toContain('SearchablePostConnection @aws_api_key @aws_iam');
  expect(out.schema).toMatchSnapshot();
});
