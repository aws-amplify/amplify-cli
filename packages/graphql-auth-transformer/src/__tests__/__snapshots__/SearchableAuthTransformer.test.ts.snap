// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test auth logic is enabled for iam/apiKey auth rules in response es resolver 1`] = `
"type Post @aws_api_key @aws_iam {
  id: ID!
  content: String
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
  secret: String @aws_iam
}

enum ModelSortDirection {
  ASC
  DESC
}

type ModelPostConnection @aws_api_key @aws_iam {
  items: [Post]!
  nextToken: String
}

input ModelStringFilterInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  contains: String
  notContains: String
  between: [String]
  beginsWith: String
}

input ModelIDFilterInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  contains: ID
  notContains: ID
  between: [ID]
  beginsWith: ID
}

input ModelIntFilterInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
}

input ModelFloatFilterInput {
  ne: Float
  eq: Float
  le: Float
  lt: Float
  ge: Float
  gt: Float
  between: [Float]
}

input ModelBooleanFilterInput {
  ne: Boolean
  eq: Boolean
}

input ModelPostFilterInput {
  id: ModelIDFilterInput
  content: ModelStringFilterInput
  secret: ModelStringFilterInput
  and: [ModelPostFilterInput]
  or: [ModelPostFilterInput]
  not: ModelPostFilterInput
}

type Query {
  searchPosts(filter: SearchablePostFilterInput, sort: SearchablePostSortInput, limit: Int, nextToken: String, from: Int): SearchablePostConnection @aws_api_key @aws_iam
  getPost(id: ID!): Post @aws_api_key @aws_iam
  listPosts(filter: ModelPostFilterInput, limit: Int, nextToken: String): ModelPostConnection @aws_api_key @aws_iam
}

input CreatePostInput {
  id: ID
  content: String
  secret: String
}

input UpdatePostInput {
  id: ID!
  content: String
  secret: String
}

input DeletePostInput {
  id: ID!
}

type Mutation {
  createPost(input: CreatePostInput!): Post @aws_api_key @aws_iam
  updatePost(input: UpdatePostInput!): Post @aws_api_key @aws_iam
  deletePost(input: DeletePostInput!): Post @aws_api_key @aws_iam
}

type Subscription {
  onCreatePost: Post @aws_subscribe(mutations: [\\"createPost\\"]) @aws_api_key @aws_iam
  onUpdatePost: Post @aws_subscribe(mutations: [\\"updatePost\\"]) @aws_api_key @aws_iam
  onDeletePost: Post @aws_subscribe(mutations: [\\"deletePost\\"]) @aws_api_key @aws_iam
}

input SearchableStringFilterInput {
  ne: String
  gt: String
  lt: String
  gte: String
  lte: String
  eq: String
  match: String
  matchPhrase: String
  matchPhrasePrefix: String
  multiMatch: String
  exists: Boolean
  wildcard: String
  regexp: String
  range: [String]
}

input SearchableIntFilterInput {
  ne: Int
  gt: Int
  lt: Int
  gte: Int
  lte: Int
  eq: Int
  range: [Int]
}

input SearchableFloatFilterInput {
  ne: Float
  gt: Float
  lt: Float
  gte: Float
  lte: Float
  eq: Float
  range: [Float]
}

input SearchableBooleanFilterInput {
  eq: Boolean
  ne: Boolean
}

input SearchableIDFilterInput {
  ne: ID
  gt: ID
  lt: ID
  gte: ID
  lte: ID
  eq: ID
  match: ID
  matchPhrase: ID
  matchPhrasePrefix: ID
  multiMatch: ID
  exists: Boolean
  wildcard: ID
  regexp: ID
  range: [ID]
}

input SearchablePostFilterInput {
  id: SearchableIDFilterInput
  content: SearchableStringFilterInput
  secret: SearchableStringFilterInput
  and: [SearchablePostFilterInput]
  or: [SearchablePostFilterInput]
  not: SearchablePostFilterInput
}

enum SearchableSortDirection {
  asc
  desc
}

enum SearchablePostSortableFields {
  id
  content
  secret
}

input SearchablePostSortInput {
  field: SearchablePostSortableFields
  direction: SearchableSortDirection
}

type SearchablePostConnection @aws_api_key @aws_iam {
  items: [Post]!
  nextToken: String
  total: Int
}
"
`;
